### DISCUSSION TITLE
Function IDs

### DISCUSSION THREAD

**Discussion by hannes** [substrait-io/substrait#64] [Q&A] [answered]
I tried mapping DuckDB query plans to substrait and it seems like scalar and aggregate functions need to be in the extension definition and are then referred to by an ID. Is this correct? 

From the documentation it also seems like those function IDs are supposed to be system-specific, is that correct, too?

**Comment by jacques-n** [comment]
Arg, totally missed this question, sorry about that. I guess it doesn't show up in the slack alerts.

Function ids themselves are just local to a particular plan and are a way of normalizing the actual function mapping which maps to a canonical location (e.g. uri) and key (just been updated to the name of the function plus the input argument types, used to be an integer offset).

The goal is for the project itself to declare a common set of frequently functions signatures but then people should be able to also define their own function uris that hold additional custom functions signatures. For example, substrait has an aggregate function `avg(i8) => i64?`. Some systems prefer to use a floating point type for average and may create another version which is `avg(i8) => fp64`. The specification allows these alternative implementations. A substrait producer would potentially identify the available set of maps that a consumer has by probing it and finding out which subset of functions it implements, including this alternative version of average.

**Reply by hannes** [reply]
Thanks Jacques for the description.

Another question, as I understand Substrait, it is supposed to be an interchange format between different query processing engines and frontends. From our SQL-centric view, it fits best into logical query plans before binding. Yet some components of Substrait seem like it is trying to be a physical plan representation as well, with types everywhere, specific function semantics (see below), parallel query processing intermediate states, things like runtime constraints etc. Which one is it or is this still up for discussion?

Some additional comments:

- Specifying the function at the current specification depth (argument types, return types, decimal casting rules, partial intermediates etc.) centrally is not going to work since it requires a lot of already non-existent consensus between implementors.
- Having system-specific extensions there creates a quadratic problem, where every new implementor of Substrait has to implement all the function definitions (or wrappers) that exist elsewhere, and in turn all existing implementors have to add support for the new guy's definitions.
- I also wonder why the function definitions use a separate encoding format (YAML), maybe it would be better to stick to one format in an already complex standard.

We propose to instead focus on centrally defining a function list with names and parameter "meanings" only. Then, there is an actual chance of query plans being usable on other systems. Yes, there is a semantics issue, but I think that cannot be solved by a plan interchange language.

**Reply by jacques-n** [reply]
> From our SQL-centric view, it fits best into logical query plans before binding. Yet some components of Substrait seem like it is trying to be a physical plan representation as well, with types everywhere, specific function semantics (see below), parallel query processing intermediate states, things like runtime constraints etc. Which one is it or is this still up for discussion?

This is a great question and the mental model that we've been developing against is definitely more Substrait-general first as opposed to specific system focused. In general, I think of things being on a logical to physical spectrum as opposed to a clear binary. For example, in many systems full type resolution/coercion/etc is done as part of validation before ever arriving at any kind of logical plan whereas other systems may defer that to much later in the process (possibly all the way to execution). Additionally, some operations have very little variance or difference in properties/behavior between their logical version and their physical version. Because the line is so blurry from my perspective, I think it would be quite difficult to say "Substrait only does physical" or "Substrait only does logical". That is why you see concepts from both ends of the spectrum in substrait and why Substrait doesn't generally focus on that distinction: it can be subjective. 

I'd additionally add that as part of the design, one of the goals around the plan was that it was explicitly clear in intention. This is relatively important in trying to ensure that the same plan can result in the same outcome when working with different systems.

Is the project open to updates/modifications of that thinking: for sure. As the community builds, our goals is to incorporate the ideas of those people. 

> Specifying the function at the current specification depth (argument types, return types, decimal casting rules, partial intermediates etc.) centrally is not going to work since it requires a lot of already non-existent consensus between implementors.
>I also wonder why the function definitions use a separate encoding format (YAML), maybe it would be better to stick to one format in an already complex standard.

These two items are actually very much connected. I don't believe there will ever be absolute consensus as there is too much that has already been built and used. For example, I believe the `substring` function has different semantics between Presto and Spark. There is enough existing usage of each of those tools that I expect these two variations of functions exist indefinitely. In Substrait, clarity of intention and semantics means that these two different functions must be expressed as two distinct function signatures. They may both be defined inside the substrait project in different yaml files but they can just easily be defined in another project. 

In other words, Substrait isn't trying to define the one `substring` to rule them all, it is trying to make sure that a plan that includes `substring` is very clear about what it means by that function name. As a point of extensibility, these lists of function semantics are something we generally think will be developed by humans. Using a human readable representation lends itself to easier creation with existing tooling. (Note that the yaml isn't the only representation, the binary protobuf also has a representation of this information for use when collaborating with systems (e.g. engine tells fe what set of functions it supports).)

> Having system-specific extensions there creates a quadratic problem, where every new implementor of Substrait has to implement all the function definitions (or wrappers) that exist elsewhere, and in turn all existing implementors have to add support for the new guy's definitions.

Yeah, it's a problem. That being said, I don't think it is a problem Substrait creates or can fix. What I believe Substrait can do is:
* Make sure we have clarity around where the problem is occurring (clear plan intention, discrete function signatures/ids)
* Provide a foundation on which a subset of systems can agree on the same set of semantics/supported operations (we all support this set of functions)
* Provide a set of lossless and lossy rewrites that help make the quadratic problem more tractable.
* Provide a standardized pattern for location of implementations in standardized forms (llvm, python, java, wasm) so that systems can actively source missing implementations.

The other note here is that I don't see being "Substrait compliant" as being the same as "you must implement all substrait defined functions". I don't think that is realistic given the set of variations that exist in the world. I do think at some point we should strive define named "sets" of functions to help with this problem preferably "postgres::8.3::scalar" or more likely "postgres::8.3::scalar::math::nondecimal" but I think it will take some time as I believe the current state is that virtually every system supports a random different subset of functions. Given this dynamic, I plan to also provide a standardized "capabilities" communication mechanism where a FE can ask a BE what it supports and it returns a subset of uri+functionnames it supports. Think of a much better/clearer version of ODBC capabilities or Tableau's tdd file.

**Reply by hannes** [reply]
Thanks Jacques for your extensive clarification on this. CCing @cpcloud  for further offline discussion.

**Reply by hannes** [reply]
As an added note, the function question is absolutely a problem Substrait creates. By insisting on precise definitions, you create a problem that you have no clear path on how to solve yet.

**Reply by cpcloud** [reply]
I think there's a bit of a disconnect here on what's needed "right now" versus what we may need in the future. Function specification is a challenging topic, clearly.

Since there are producers and consumers in active development with needs met by a much simpler function specification, how about we simplify the specification until there are producers/consumers that actually meet the criteria for needing something with the current level of complexity.

I believe that we should onboard as many systems as reasonably possible and get people building things ASAP to validate the specification. Currently, the people building those things are finding it rather challenging to handle function calls in a compliant way.

For example, I'm not sure we have a producer under active development that is distributing partial aggregate plans to N possibly different consumers, while we do have multiple producer/consumer pairs that will not need this functionality for quite awhile (possibly ever).

Since Substrait is a new effort with a lot of value to add, I think we should focus on getting developers on-boarded to Substrait as quickly as possible.

Without doing this, I think that:
1. developers will simply ignore the function spec because they don't need it for their use case. This might be fine for the use case, but certainly limits the utility of Substrait and leads to a bunch of non-spec-compliant code.
2. developers will not use Substrait, because while it may solve a problem they have, it's too difficult to adhere to the function spec with their existing system. @jacques-n as you said, there are already many existing systems that add a lot of value and behave in ways that cannot be reasonably changed.

I think we have to make it easier for existing systems to adhere to substrait, which I believe means we need to simplify the function spec and add depth as the need arises.

**Reply by westonpace** [reply]
I think I am confused on the perceived burden here.

> how about we simplify the specification until there are producers/consumers that actually meet the criteria for needing something with the current level of complexity.

Given the simplification we have already done there is no need for a producer or consumer to read the functions yaml so I'm not sure what could be simplified further.

> As an added note, the function question is absolutely a problem Substrait creates. By insisting on precise definitions, you create a problem that you have no clear path on how to solve yet.

If a system doesn't want to solve that problem then there is an alternative.  Specify all functions as belonging to your own extension.  There is no need to actually write a YAML for it.  Just make up some URL and then as long as your producers and consumers internally agree on the meaning of the names you should be fine.

When you want interoperability with other systems then it seems that there would need to be some level of agreement on the meaning behind those names.

> developers will simply ignore the function spec because they don't need it for their use case. This might be fine for the use case, but certainly limits the utility of Substrait and leads to a bunch of non-spec-compliant code.

I guess I fall into this camp.  I think it's fine.  You said it yourself that many people don't need this yet.  People will not want to do the work for interoperability until there is a need for it.  Once there is a need they can do the work to be compliant.  The functions.yaml isn't really going to lead to spec-compliant code by itself anyways.  You would need a suite of integration tests to truly get to something you can trust.

> I think we have to make it easier for existing systems to adhere to substrait, which I believe means we need to simplify the function spec and add depth as the need arises.

> That said, I do think it is possible to at least agree on function names, arity, and perhaps even argument and return types.

Isn't that exactly what the function spec is currently doing?  What could be simplified out?

**Reply by jacques-n** [reply]
I think there is some ambiguity around the current specification. This is probably largely due to weak documentation and bad naming. As such, I've tried to address both in #76. By and large, the capability you appear to seek already exists in the specification. To make it clearer, I've posted a patch to provides additional documentation as well as an example [unknown.yaml](https://github.com/substrait-io/substrait/blob/cde8f765d314538ec07972d80a899f90eba2eb7a/extensions/unknown.yaml) extension file that exposes how one could express a simple set of functions that are type opaque. 

I've also added specific clarifications in the function name mapping specification that clarifies that any function that is a single-impl function can be referenced using the simple name (without any argument type additions):
> If there is only one impl, the name may be defined at the top-level definition of the object. If there are multiple impls, the name is a composite string that includes the top-level name (e.g. add) followed by a compound string that describes the types of the arguments (e.g. add:i8_i8). In the case there is one impl, the multiple impl naming scheme is also supported.

**Comment by westonpace** [comment]
Note that missing functionality can often be supplied by plan transformations (sometimes at a cost to performance) as well.  For example, system A might support "add with saturating overflows" and system B may only support "add with null overflows".  If the producer requests "add with saturating overflows" then a transformation library could write a plan for system B that does a coalesce after the add.

There's also been some discussion of "optional behavior".  For example, a producer might say "add and I don't care how overflow is handled" and both system A and system B could consume the plan and the producer has acknowledged and clearly communicated the behavior may depend on the system.

An analogue might be something like instruction sets.  x86 is a well known instruction set that many processors support.  There are standard extensions like SSE4.2 that are multi-vendor but only provided by newer or more extensive processors.  And there are extensions like FMA4 which are, at the moment, single-vendor only.

**Reply by cpcloud** [reply]
I do think the instruction sets analogy could work, but personally it would make me sad if this is how Substrait evolved.

The problem with instruction sets is that effectively a single vendor owned the specification. I think we're explicitly trying to avoid such a scenario and get more "vendors" (keeping the terminology from the analogy) involved so that one voice doesn't dominate the outcome.

That said, I do think it is possible to at _least_ agree on function names, arity, and perhaps even argument and return types.

**Comment by jacques-n** [comment, marked as answer]
The documentation was updated in #76 and people will hopefully find answers in the [extensions](https://substrait.io/extensions/) docs as well as the [binary serialization](https://substrait.io/serialization/binary_serialization/) docs.
